thrifty-runtime/src/main/java/com/microsoft/thrifty/service/ClientBase.java

```
@@ -85,226 +48,96 @@
-
-            protocol.writeMessageBegin(call.name, call.callTypeId, sid);
-            call.send(protocol);
-            protocol.writeMessageEnd();
-            protocol.flush();
-
-            if (isOneWay) {
-                // No response will be received
-                complete(call, null);
-                return;
-            }
-
-            MessageMetadata metadata = protocol.readMessageBegin();
-
-            if (metadata.seqId != sid) {
-                throw new ThriftException(
-                        ThriftException.Kind.BAD_SEQUENCE_ID,
-                        "Unrecognized sequence ID");
-            }
-
-            if (metadata.type == TMessageType.EXCEPTION) {
-                ThriftException e = ThriftException.read(protocol);
-                fail(call, e);
-                protocol.readMessageEnd();
-            } else if (metadata.type != TMessageType.REPLY) {
-                throw new ThriftException(
-                        ThriftException.Kind.INVALID_MESSAGE_TYPE,
-                        "Invalid message type: " + metadata.type);
-            }
-
-            if (metadata.seqId != seqId.get()) {
-                throw new ThriftException(
-                        ThriftException.Kind.BAD_SEQUENCE_ID,
-                        "Out-of-order response");
-            }
-
-            if (!metadata.name.equals(call.name)) {
-                throw new ThriftException(
-                        ThriftException.Kind.WRONG_METHOD_NAME,
-                        "Unexpected method name in reply; expected " + call.name
-                                + " but received " + metadata.name);
-            }
+        if (metadata.type == TMessageType.EXCEPTION) {
+            ThriftException e = ThriftException.read(protocol);
+            protocol.readMessageEnd();
+            handleExceptionMessage(call, e);
+            //TODO there was no throw/return here before but the sync impl of handleExceptionMessage throws
```

I've changed this to throw a custom exception after `handleExceptionMessage()`. For sync that will never be reached because the `ThriftException` is thrown before. The async client will catch the custom exception and just swallow it because it's `handleExceptionMessage()` already called `fail()`.  (Kotlin's `Nothing` as return type for `handleExceptionMessage()` would be really nice here)

We should probably add a test that the async client doesn't call `fail()` twice after receiving a `TMessageType.EXCEPTION`. I'm not very familiar with server thrift code, is there an easy way to make it send one?